src/server/lock_manager/waiter_manager.rs

```diff
@@ -220,179 +247,155 @@ impl Waiter {
-            },
-            ProcessResult::Failed { err } => match err {
-                StorageError(box StorageErrorInner::Txn(TxnError(box TxnErrorInner::Mvcc(
-                    MvccError(box MvccErrorInner::WriteConflict {
-                        ref mut key,
-                        ref mut primary,
-                        ..
-                    }),
-                )))) => (std::mem::take(key), std::mem::take(primary)),
-                _ => panic!("unexpected mvcc error"),
-            },
-            _ => panic!("unexpected progress result"),
-        }
+        self.cancel(Some(StorageError::from(TxnError::from(e))))
     }
 }
 
-// NOTE: Now we assume `Waiters` is not very long.
-// Maybe needs to use `BinaryHeap` or sorted `VecDeque` instead.
-type Waiters = Vec<Waiter>;
-
 struct WaitTable {
-    // Map lock hash to waiters.
-    wait_table: HashMap<u64, Waiters>,
+    // Map lock hash and ts to waiters.
+    // For compatibility.
+    wait_table: HashMap<(u64, TimeStamp), LockWaitToken>,
+    waiter_pool: HashMap<LockWaitToken, Waiter>,
     waiter_count: Arc<AtomicUsize>,
+
+    wake_up_key_delay_callback: Option<Box<dyn Fn(&Key, TimeStamp, TimeStamp, Instant) + Send>>,
 }
 
 impl WaitTable {
     fn new(waiter_count: Arc<AtomicUsize>) -> Self {
         Self {
             wait_table: HashMap::default(),
+            waiter_pool: HashMap::default(),
             waiter_count,
+            wake_up_key_delay_callback: None,
         }
     }
 
+    fn set_wake_up_key_delay_callback(
+        &mut self,
+        cb: Option<Box<dyn Fn(&Key, TimeStamp, TimeStamp, Instant) + Send>>,
+    ) {
+        self.wake_up_key_delay_callback = cb;
+    }
+
     #[cfg(test)]
     fn count(&self) -> usize {
-        self.wait_table.iter().map(|(_, v)| v.len()).sum()
+        self.waiter_pool.len()
     }
 
     fn is_empty(&self) -> bool {
-        self.wait_table.is_empty()
+        self.waiter_pool.is_empty()
     }
 
     /// Returns the duplicated `Waiter` if there is.
-    fn add_waiter(&mut self, waiter: Waiter) -> Option<Waiter> {
-        let waiters = self.wait_table.entry(waiter.lock.hash).or_insert_with(|| {
-            WAIT_TABLE_STATUS_GAUGE.locks.inc();
-            Waiters::default()
-        });
-        let old_idx = waiters.iter().position(|w| w.start_ts == waiter.start_ts);
-        waiters.push(waiter);
-        if let Some(old_idx) = old_idx {
-            let old = waiters.swap_remove(old_idx);
-            self.waiter_count.fetch_sub(1, Ordering::SeqCst);
-            Some(old)
-        } else {
-            WAIT_TABLE_STATUS_GAUGE.txns.inc();
-            None
-        }
-        // Here we don't increase waiter_count because it's already updated in
-        // LockManager::wait_for()
-    }
+    fn add_waiter(&mut self, token: LockWaitToken, waiter: Waiter) -> bool {
+        self.wait_table
+            .insert((waiter.wait_info.lock_digest.hash, waiter.start_ts), token);
```

Comment:
Indexing by (hash, start_ts) is here to provide compatibility since the deadlock detector tries to wake up by (hash, start_ts) when it founds deadlock. I only supposed to keep the token unique, and just replace and ignore it when there are duplicated (hash, start_ts) pair. May there be any problem?

Reply:
OK. Now I see the duplicated waiter token will finally be cleared when it times out.

It seems not a big problem that the old duplicated waiter is changed to get notified at timeout instead of immediately.